File: server/internal/sentry/sentry.go
package sentry

import (
	"log"
	"os"
	"time"

	"github.com/getsentry/sentry-go"
)

// Initialize sets up Sentry for the Go server
func Initialize() error {
	dsn := os.Getenv("SENTRY_GO_DSN")
	if dsn == "" {
		log.Println("[Sentry] Go backend DSN not configured, skipping initialization")
		return nil
	}

	err := sentry.Init(sentry.ClientOptions{
		Dsn:              dsn,
		Environment:      getEnvironment(),
		Release:          os.Getenv("SENTRY_RELEASE"),
		EnableLogs:       true, // Enable logs to be sent to Sentry
		TracesSampleRate: 1.0,  // Capture 100% of transactions for tracing
		AttachStacktrace: true,
		BeforeSend: func(event *sentry.Event, hint *sentry.EventHint) *sentry.Event {
			// Filter out development errors unless explicitly enabled
			if os.Getenv("GO_ENV") != "production" && os.Getenv("SENTRY_DEBUG") != "true" {
				return nil
			}
			return event
		},
	})

	if err != nil {
		return err
	}

	log.Println("[Sentry] Go backend initialized successfully")
	return nil
}

// Flush ensures all events are sent before shutdown
func Flush() {
	sentry.Flush(2 * time.Second)
}

// CaptureError sends an error to Sentry
func CaptureError(err error, tags map[string]string) {
	if err == nil {
		return
	}

	sentry.WithScope(func(scope *sentry.Scope) {
		for key, value := range tags {
			scope.SetTag(key, value)
		}
		sentry.CaptureException(err)
	})
}

// CaptureMessage sends a message to Sentry
func CaptureMessage(message string, level sentry.Level) {
	sentry.CaptureMessage(message)
}

// RecoverPanic recovers from panics and sends to Sentry
func RecoverPanic() {
	if r := recover(); r != nil {
		sentry.CurrentHub().Recover(r)
		sentry.Flush(2 * time.Second)
		panic(r) // Re-panic after capturing
	}
}

func getEnvironment() string {
	env := os.Getenv("GO_ENV")
	if env == "" {
		return "development"
	}
	return env
}

File: web-legacy/src/server/sentry.ts
/**
 * Sentry Configuration for Bun Backend Server
 *
 * Uses @sentry/bun for optimal Bun runtime integration
 * Separate from frontend Sentry for better error tracking and distributed tracing
 */

import * as Sentry from '@sentry/bun';

/**
 * Initialize Sentry for the Bun backend server
 */
export function initServerSentry() {
  const dsn = process.env.SENTRY_SERVER_DSN || '';

  // Skip initialization if no DSN provided
  if (!dsn) {
    console.log('[Sentry] Backend DSN not configured, skipping initialization');
    return;
  }

  Sentry.init({
    // Use separate DSN for backend project
    dsn: dsn,
    environment: process.env.NODE_ENV || 'development',
    release: process.env.SENTRY_RELEASE || 'dev',

    // Send structured logs to Sentry
    enableLogs: true,

    // Enable distributed tracing
    tracesSampleRate: 1.0, // Capture 100% of the transactions

    // Bun-specific integrations
    integrations: [
      // Bun's built-in HTTP server instrumentation
      Sentry.bunServerIntegration(),

      // Automatically capture console errors
      Sentry.captureConsoleIntegration({
        levels: ['error', 'warn'],
      }),
    ],

    // Filter out noisy errors
    beforeSend(event, hint) {
      // Skip development errors unless explicitly enabled
      if (process.env.NODE_ENV !== 'production' && !process.env.SENTRY_DEBUG) {
        return null;
      }

      const error = hint.originalException;

      // Filter out common non-critical errors
      if (error && typeof error === 'object' && 'message' in error) {
        const message = (error as Error).message;

        // Skip connection errors in development
        if (process.env.NODE_ENV !== 'production') {
          if (message.includes('ECONNREFUSED') || message.includes('ENOTFOUND')) {
            return null;
          }
        }
      }

      return event;
    },

    // Add custom context
    beforeBreadcrumb(breadcrumb) {
      // Filter out verbose HTTP breadcrumbs
      if (breadcrumb.category === 'http' && breadcrumb.data?.method === 'GET') {
        // Only keep failed requests
        if (breadcrumb.data?.status_code && breadcrumb.data.status_code < 400) {
          return null;
        }
      }

      return breadcrumb;
    },
  });

  console.log('[Sentry] Backend initialized successfully');
}

/**
 * Bun.serve() integration
 *
 * For Bun's native server, Sentry.bunServerIntegration() automatically instruments
 * fetch handlers. No additional middleware is needed.
 *
 * To capture errors manually in your Bun server:
 * - Use Sentry.captureException(error) for errors
 * - Use Sentry.captureMessage(message) for messages
 * - Console errors/warnings are automatically captured if captureConsoleIntegration is enabled
 */

// Export Sentry for use in other modules
export { Sentry };
